Don't allow backups to be made via schedules if limit = 0

notAreYouScared opened the pull request:
Closes https://github.com/pterodactyl/panel/issues/2395

Original PR https://github.com/pterodactyl/panel/pull/2416

~~I have no idea how to protect the endpoint to prevent the task from being made via the API :(~~

API is protected 👍 

Also, fixes the title when directly navigating to the details of a specific schedule. ( panel.com/server/<uuid>/schedules/# )

notAreYouScared committed:
Don't allow backups to be made via schedules if limit = 0

Closes https://github.com/pterodactyl/panel/issues/2395

Original PR https://github.com/pterodactyl/panel/pull/2416

I have no idea how to protect the endpoint to prevent the task from being made via the api :(

notAreYouScared committed:
Update TaskDetailsModal.tsx

nikrozman21 commented:
This looks like a useful feature to me, but you are probably missing a column in the schedules migration.

notAreYouScared commented:
> This looks like a useful feature to me, but you are probably missing a column in the schedules migration.

Huh? I'm not adding or editing any migrations? What's missing?

nikrozman21 commented:
> > This looks like a useful feature to me, but you are probably missing a column in the schedules migration.
> 
> Huh? I'm not adding or editing any migrations? What's missing?

Sorry, accidentally posted it on the wrong thread.
Comment hidden.

notAreYouScared committed:
Protect API

Protects API from creating/updating tasks when backup limit = 0

Also adds click to copy for client API key when modal is visible.

notAreYouScared committed:
.....

Match the rest of the tabs <3

matthewpi commented on app/Http/Controllers/Api/Client/Servers/ScheduleTaskController.php:
```diff
@@ -49,6 +50,10 @@ public function store(StoreTaskRequest $request, Server $server, Schedule $sched
         }
 
+        if ($server->backup_limit === 0 && $request->action === 'backup'){
+            throw new HttpForbiddenException("A backup task cannot be created when the backup limit for a server is set to {$server->backup_limit}.");
```
This should return a bad request response, forbidden is used for bad tokens.  Message should be changed to something like `This server is not able to create any backups` or `A backup task cannot be created because the server has a backup limit of 0` (if you leave the message the same, you can replace `{$server->backup_limit}` with `0`).

notAreYouScared replied:
![image](https://user-images.githubusercontent.com/1757840/117086853-c2de4180-ad1b-11eb-9d2b-72ac3fb145dd.png)
![image](https://user-images.githubusercontent.com/1757840/117086857-c671c880-ad1b-11eb-8d25-82af6731232a.png)

matthewpi commented on app/Http/Controllers/Api/Client/Servers/ScheduleTaskController.php:
```diff
@@ -49,6 +50,10 @@ public function store(StoreTaskRequest $request, Server $server, Schedule $sched
             throw new ServiceLimitExceededException("Schedules may not have more than {$limit} tasks associated with them. Creating this task would put this schedule over the limit.");
         }
 
+        if ($server->backup_limit === 0 && $request->action === 'backup'){
```
`){` -> `) {`

matthewpi commented on app/Http/Controllers/Api/Client/Servers/ScheduleTaskController.php:
```diff
@@ -81,6 +87,10 @@ public function update(StoreTaskRequest $request, Server $server, Schedule $sche
             throw new NotFoundHttpException();
         }
 
+        if ($server->backup_limit === 0 && $request->action === 'backup'){
```
`){` -> `) {`

matthewpi commented on app/Http/Controllers/Api/Client/Servers/ScheduleTaskController.php:
```diff
@@ -81,6 +87,10 @@ public function update(StoreTaskRequest $request, Server $server, Schedule $sche
         }
 
+        if ($server->backup_limit === 0 && $request->action === 'backup'){
+            throw new HttpForbiddenException("A backup task cannot be created when the backup limit for a server is set to {$server->backup_limit}.");
```
This should return a bad request response, forbidden is used for bad tokens.  Message should be changed to something like `This server is not able to create any backups` or `A backup task cannot be created because the server has a backup limit of 0` (if you leave the message the same, you can replace `{$server->backup_limit}` with `0`).

notAreYouScared replied:
Changed to just 0 as the message would only be shown if the limit === 0. 
as for the wording, it's from Dane...

matthewpi commented on resources/scripts/components/server/schedules/TaskDetailsModal.tsx:
```diff
@@ -78,21 +79,27 @@ const TaskDetailsModal = ({ schedule, task }: Props) => {
-                    tasks = [ ...tasks, task ];
-                }
+        if (backupLimit === 0 && values.action === 'backup') {
+            console.error('This server is not allowed to create backups.');
```
You don't need the `console.error` if you are showing a flash to the user.

notAreYouScared replied:
I was just following what was above it... Should it be removed in both places?

notAreYouScared replied:
Removed.

matthewpi commented on resources/scripts/components/server/schedules/TaskDetailsModal.tsx:
```diff
@@ -78,21 +79,27 @@ const TaskDetailsModal = ({ schedule, task }: Props) => {
+        if (backupLimit === 0 && values.action === 'backup') {
+            console.error('This server is not allowed to create backups.');
+            setSubmitting(false);
+            addError({ message: 'This server is not allowed to create backups.', key: 'schedule:task' });
```
This message might be better being `This server is not able to create backups` rather than `not allowed`.

notAreYouScared replied:
Changed

notAreYouScared committed:
Attempt 4

@matthewpi Changes applied.

notAreYouScared commented:
Whelp I broke the schedule creation test, as it fails on making a backup... Guess that means I blocked it correctly :P

I would add 
```PHP

    /**
     * Test that backups can not be tasked when the backup limit is 0
     */
    public function testBackupsCanNotBeTaskedIfLimit0()
    {
        [$user, $server] = $this->generateTestAccount();

        /** @var \Pterodactyl\Models\Schedule $schedule */
        $schedule = Schedule::factory()->create(['server_id' => $server->id]);

        $this->actingAs($user)->postJson($this->link($schedule, '/tasks'), [
            'action' => 'backup',
            'time_offset' => 0,
        ])
            ->assertStatus(Response::HTTP_FORBIDDEN)
            ->assertJsonPath('errors.0.detail', 'A backup task cannot be created when the server\'s backup limit is set to 0.');

        $this->actingAs($user)->postJson($this->link($schedule, '/tasks'), [
            'action' => 'backup',
            'payload' => "file.txt\nfile2.log",
            'time_offset' => 0,
        ])
            ->assertStatus(Response::HTTP_FORBIDDEN)
            ->assertJsonPath('errors.0.detail', 'A backup task cannot be created when the server\'s backup limit is set to 0.');
    }
 ```    
So it would have a test that would pass if backup_limit = 0
I don't know how to set the variable so the backup_limit would be >1

notAreYouScared committed:
Properly update test to expect it to fail.

Fixes test so it expects it to fail.

DaneEveritt committed:
Merge branch 'develop' into backupthingy
